### PR TITLE
Fix TransactionPool

### PR DESCRIPTION
**Pull request by marsupial**
Transaction::~Transaction being called twice.
Remove extra Transaction allocation/leak per Transaction.

### CODE REVIEW

File: lib/Interpreter/TransactionPool.h [hunk 1]
```diff
@@ -20,74 +20,59 @@ namespace clang {
...
 
-      T->m_State = Transaction::kCollecting;
       return T;
     }
 
-    void releaseTransaction(Transaction* T) {
-      assert((T->getState() == Transaction::kCompleted ||
-              T->getState() == Transaction::kRolledBack)
-             && "Transaction must completed!");
+    void releaseTransaction(Transaction* T, bool reuse = true) {
```
Comment (karies) [inline lib/Interpreter/TransactionPool.h]: The original code had an `#ifndef NDEBUG`, to force allocation of new transactions. Can we keep this in some way, maybe setting `reuse` to false?

Comment (marsupial) [inline lib/Interpreter/TransactionPool.h]: I can.
Out of curiosity what is the reasoning?

Comment (marsupial) [inline lib/Interpreter/TransactionPool.h]: For the comment.

Comment (karies) [inline lib/Interpreter/TransactionPool.h]: We cat set watch points if transaction number N always ends up at a given address. If they are re-used then the transaction N is at the same memory address as a gazillion transactions before. That makes watch points much more difficult to use...

Comment (marsupial) [inline lib/Interpreter/TransactionPool.h]: Added.

File: lib/Interpreter/TransactionPool.h [hunk 2]
```diff
@@ -52,7 +57,7 @@ namespace cling {
       return T;
     }
 
-    void releaseTransaction(Transaction* T, bool reuse = true) {
+    void releaseTransaction(Transaction* T, bool reuse = !kDebugMode) {
```
Comment (karies) [inline lib/Interpreter/TransactionPool.h]: For my brain, that makes the logic more complex than needed: this setting should override the value passed as `reuse`  if `reuse` is `true` and `NDEBUG` is off. I.e. in `#ifndef NDEBUG`, `reuse` shall be `false`, whatever the argument passed. With your patch I'll need to check that no caller calls it with an explicit `true`, now or in the future.

Can we "hard-wire" this by overriding the value of `reuse = false` in `releaseTransaction`'s body for `#ifndef NDEBUG`?

Comment (marsupial) [inline lib/Interpreter/TransactionPool.h]: Ok?

**Comment by karies** [comment]
Perfect, thanks! Merged!

**Pull request closed without merging**